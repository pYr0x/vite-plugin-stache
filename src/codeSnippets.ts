import path from "path";

export function createCodeSnippetForDynamicImports(dynamicImports: string[]): string {
  if(dynamicImports.length) {
    return `const dynamicImportMap = Object.assign({}, ${dynamicImports.map((file) => {
      if(!path.extname(file)){
        file += '.js';
      }
      return `import.meta.glob('${file}')`;
    }).join(",")});
            dynamicImporter(dynamicImportMap);`;
  }
  return '';
}

export function createCodeSnippetForStaticImports(staticImports: string[]): string {
  if(staticImports.length) {
    return `const staticImportMap = [${staticImports.map((file) => {
      return `"${file}"`;
    }).join(",")}];
            staticImporter(staticImportMap);`;
  }
   return '';
}
