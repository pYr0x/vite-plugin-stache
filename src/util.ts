export function identifyImports(imports: string[], importDeclarations: ImportDeclaration[]) {
  const tagImportMap: string[] = [];
  const simpleImports: string[] = [];

  imports.forEach((file) => {
    for (let importFile of importDeclarations) {
      if (importFile && importFile.specifier === file && importFile.attributes instanceof Map) {
        if(importFile.attributes.size > 1) {
          tagImportMap.push(importFile.specifier);
          break;
        }else if(importFile.attributes.size === 1){
          simpleImports.push(importFile.specifier);
          break;
        }
      }
    }
  });
  return {simpleImports, tagImportMap};
}
