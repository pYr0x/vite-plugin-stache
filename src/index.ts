import {parse} from "can-stache-ast";
// const makeSourceMap = require( "./source-map" );
import {Plugin} from 'vite'
import process from "process";
import stacheTransformer from "stache-inline-transformer";
import whichModules from "stache-inline-transformer/dist/transformer/modules";
import {createCodeSnippetForDynamicImports, createCodeSnippetForStaticImports} from "./codeSnippets";
import {Options, ResolvedOptions} from "../types/plugin";
import {identifyImports} from "./util";


function stacheFilePlugin(rawOptions: Options = {}): Plugin {
  const {
    include = /\.stache$/,
    exclude
  } = rawOptions

  let options: ResolvedOptions = {
    isProduction: process.env.NODE_ENV === 'production',
    ...rawOptions,
    include,
    exclude,
    root: process.cwd()
  }
  return {
    name: 'vite:stache',
    configResolved(config) {
      options = {
        ...options,
        root: config.root,
        sourceMap: config.command === 'build' ? !!config.build.sourcemap : true,
        isProduction: config.isProduction
      }
    },
    transform(code: string, id: string) {
      const [filename] = id.split('?', 2)
      code = code.trim()
      if (filename.endsWith('.stache') && !code.startsWith('import')) {

        const ast = parse(id, code);
        const intermediate = JSON.stringify(ast.intermediate);

        const {simpleImports, tagImportMap} = identifyImports( [...new Set(ast.imports)], ast.importDeclarations);

        const dynamicImports = ast.dynamicImports || [];
        // console.log(ast);

        // language=JavaScript
        const body = `
          import stache from 'can-stache';
          import Scope from 'can-view-scope';
          import 'can-view-import';
          import 'can-stache/src/mustache_core';
          import stacheBindings from 'can-stache-bindings';

          ${(Object.keys(dynamicImports).length || tagImportMap.length || simpleImports.length) ? `import {staticImporter, dynamicImporter} from 'vite-stache-import-module';`: ``}
          ${tagImportMap.map((file, i) => `import * as i_${i} from '${file}';`).join('\n')}
          ${simpleImports.map((file) => `import '${file}';`).join('\n')}

          ${(tagImportMap.length || simpleImports.length) && createCodeSnippetForStaticImports([...tagImportMap, ...simpleImports])}

          stache.addBindings(stacheBindings);
          var renderer = stache(${intermediate});

          ${dynamicImports.length && createCodeSnippetForDynamicImports(dynamicImports)}

          export default function (scope, options, nodeList) {
            if (!(scope instanceof Scope)) {
              scope = new Scope(scope);
            }
            var variableScope = scope.getScope(function (s) {
              return s._meta.variable === true
            });
            if (!variableScope) {
              scope = scope.addLetContext();
              variableScope = scope;
            }
            var moduleOptions = Object.assign({}, options);
            Object.assign(variableScope._context, {
              module: null,
              tagImportMap: {${tagImportMap.map((file, i) => `"${file}": i_${i}`).join(',')}}
            });

            return renderer(scope, moduleOptions, nodeList);
          };`;

        // const sourceMap = makeSourceMap( body, code.trim(), filename )

        return {
          code: body,
          map: {mappings: ''}
          // map: sourceMap
        };
      }
      return null;
    }
  }
}

function stacheImportPlugin(rawOptions: Options = {}): Plugin {
  let options: ResolvedOptions = {
    isProduction: process.env.NODE_ENV === 'production',
    ...rawOptions,
    root: process.cwd()
  }

  return {
    name: 'vite:stache-import-module',
    configResolved(config) {
      options = {
        ...options,
        root: config.root,
        sourceMap: config.command === 'build' ? !!config.build.sourcemap : true,
        isProduction: config.isProduction
      }
    },
    resolveId ( source ) {
      if (source === 'vite-stache-import-module') {
        return source;
      }
      return null;
    },
    load ( id ) {
      if (id === 'vite-stache-import-module') {
        // language=JavaScript
        return `
          import {flushLoader, addLoader} from 'can-import-module';

          flushLoader();

          export function staticImporter(staticImportMap){
            addLoader((moduleName) => {
              if (staticImportMap.indexOf(moduleName) !== -1) {
                return Promise.resolve();
              }
            });
          }

          export function dynamicImporter(dynamicImportMap) {
            addLoader((moduleName) => {
              if (!(moduleName.match(/[^\\\\\\\/]\\.([^.\\\\\\\/]+)$/) || [null]).pop()) {
                moduleName += '.js';
              }
              if (moduleName in dynamicImportMap) {
                return dynamicImportMap[moduleName]()
              }
            });
          }`
      }
      return null;
    }
  };
}

function stacheInlinePlugin(rawOptions: Options = {}): Plugin {
  let options: ResolvedOptions = {
    isProduction: process.env.NODE_ENV === 'production',
    ...rawOptions,
    root: process.cwd()
  }
  return {
    name: 'vite:stache-inline',
    configResolved(config) {
      options = {
        ...options,
        root: config.root,
        sourceMap: config.command === 'build' ? !!config.build.sourcemap : true,
        isProduction: config.isProduction
      }
    },
    transform(code: string, id: string) {
      const [filename] = id.split('?', 2)

      if (filename.endsWith('.js') && /node_modules/.exec(id) === null) {
        if (whichModules(code).length > 0) {
          const newCode = stacheTransformer(code);
          return {
            code: newCode,
            map: {mappings: ''}
          };
        }
      }
      return null;
    }
  }
}

export default function stachePlugin(rawOptions: Options = {}): Plugin[]{
  const {
    inlineTransformation = true,
  } = rawOptions
  const plugins: Plugin[] = [
    stacheFilePlugin(rawOptions),
    stacheImportPlugin(rawOptions)
  ];

  if(inlineTransformation){
    plugins.push(stacheInlinePlugin(rawOptions))
  }
  return plugins;
}
// // overwrite for cjs require('...')() usage
// module.exports = stachePlugin
// stachePlugin['default'] = stachePlugin
