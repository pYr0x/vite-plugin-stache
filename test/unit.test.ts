// import {assert, describe, expect, it, suite, test} from 'vitest'
import {describe, expect, it, should} from 'vitest'
import stachePlugin from "../src";
import {createCodeSnippetForDynamicImports, createCodeSnippetForStaticImports} from "../src/codeSnippets";
import {identifyImports} from "../src/util";

should();

interface ImportDeclaration {
  specifier: string,
  loc: {
    line: number
  },
  attributes: Map<string, string>
}


describe.concurrent('Plugin usage', () => {
  it('should be able to use the plugin', () => {
    expect(stachePlugin).to.be.a('function');
  });
  it('should have inlineTransformer in plugin array by default', () => {
    const plugins = stachePlugin();
    // last item in array should be inlineTransformer
    expect(plugins[plugins.length - 1].name).to.equal('vite:stache-inline');
  });
  it('disable inlineTransformer', () => {
    const plugins = stachePlugin({inlineTransformation: false});
    plugins.should.have.length(2)
    expect(plugins[plugins.length - 1].name).not.to.equal('vite:stache-inline');
  });
})
describe.concurrent('stache file', () => {

  it('should skip tranforming if file is not a stache file', async() => {
    const [filePlugin, , ] = stachePlugin();
    const transform = filePlugin.transform('var sum = 1+1;', 'test.js');
    expect(transform).to.be.null;
  });

  // stacheFilePlugin should return a object if filename ends with .stache
  it('should return a object with a code element if filename ends with .stache', async() => {
    const [filePlugin,] = stachePlugin();
    const transform = filePlugin.transform('', 'test.stache');
    expect(transform).to.be.a('object');
    expect(transform).to.have.property('code');
  });

  it('create a code snippet for dynamic imported files and adds .js extension to all files they havenÂ´t', async() => {
    const codeSnippet = createCodeSnippetForDynamicImports(['./foo' ]);
    expect(codeSnippet).toMatchSnapshot();
    // search for foo.js in codeSnippet
    expect(codeSnippet).to.contain('foo.js');
  });

  it('create a code snippet for static imported files', async() => {
    const codeSnippet = createCodeSnippetForStaticImports(['./foo' ]);
    expect(codeSnippet).toMatchSnapshot();
    // search for foo in codeSnippet
    expect(codeSnippet).to.contain('foo');
  });

});

describe.concurrent('stache import', () => {
  // stacheImportPlugin load function should return a string if the id is vite-stache-import-module
  it('should return a string if the id is vite-stache-import-module', async() => {
    const [, importPlugin] = stachePlugin();
    const id = 'vite-stache-import-module';
    const load = importPlugin.load(id);
    expect(load).to.be.a('string');
  });
  // stacheImportPlugin load function should return null if id is not vite-stache-import-module
  it('should return null if id is not vite-stache-import-module', async() => {
    const [, importPlugin] = stachePlugin();
    const id = 'any-other-imported-module';
    const load = importPlugin.load(id);
    expect(load).to.be.null;
  });
});

describe.concurrent('stache inline', () => {
  // stacheInlinePlugin do not transform if file is located in node_modules
  it('should not transform if file is located in node_modules', async() => {
    const [, , inlinePlugin] = stachePlugin();
    const transform = inlinePlugin.transform('', 'node_modules/foo/bar.js');
    expect(transform).to.be.null
  });
  // stacheInlinePlugin do not transform the code if does not machtes the stache pattern
  it('should not transform the code if does not machtes the stache pattern', async() => {
    const [, , inlinePlugin] = stachePlugin();
    const transform = inlinePlugin.transform(`var sum = 1+1;`, 'test.js');
    expect(transform).to.be.null
  });
  // stacheInlinePlugin pass the code through the inlineTransformer if the stache pattern: StacheElement is found
  it('should pass the javascript code through the inlineTransformer if the stache pattern: StacheElement is found ', async() => {
    const [, , inlinePlugin] = stachePlugin();
    const transform = inlinePlugin.transform(`class Counter extends StacheElement {}`, 'test.js');
    expect(transform).to.be.a('object');
    expect(transform).to.have.property('code');
  });
});

describe.concurrent('utils', () => {
  it('identifyImports split correctly in simple imports and taged imports', async() => {
    const imports = ['./foo', './bar'];


    const importDeclarations: ImportDeclaration[] = [
      {
        "specifier": "./foo",
        "loc": {
          "line": 1
        },
        "attributes": new Map([["form", "./foo"], ["value:to", "scope.vars.foo"]])
      },
      {
        "specifier": "./bar",
        "loc": {
          "line": 2
        },
        "attributes": new Map([["form", "./bar"]])
      }
    ];
    const {simpleImports, tagImportMap} = identifyImports(imports, importDeclarations);
    expect(simpleImports).to.deep.equal(['./bar']);
    expect(tagImportMap).to.deep.equal(['./foo']);
  });
});

