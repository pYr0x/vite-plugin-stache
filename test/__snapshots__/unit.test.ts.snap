// Vitest Snapshot v1

exports[`stache file > create a code snippet for static imported files 1`] = `
"const dynamicImportMap = Object.assign({}, import.meta.glob('./foo.js'));
            dynamicImporter(dynamicImportMap);"
`;

exports[`stache file > create a code snippet for static imported files 2`] = `
"const staticImportMap = [\\"./foo\\"];
            staticImporter(staticImportMap);"
`;
